Sources/NIOCore/Docs.docc/swift-concurrency.md

```diff
@@ -0,0 +1,244 @@
+need to be known. Moreover, these types are not allowed to change anymore otherwise it will result in
+runtime crashes.
+To solve the problem of protocol negotiation we had to introduce a new ``ChannelHandler`` protocol called
+`NIOProtocolNegotiationHandler`. This protocol requires a single future property `NIOProtocolNegotiationHandler/protocolNegotiationResult`
+that is completed once the handler is finished with protocol negotiation. In the successful case,
+the future can either indicate that protocol negotiation is fully done by returning `NIOProtocolNegotiationResult/finished(_:)``or
+indicate that further protocol negotiation needs to be done by returning `NIOProtocolNegotiationResult/deferredResult(_:)`.
+
+#### ServerBootstrap
+
+The server bootstrap is used to create a new TCP based server. Once any of the bind methods on the `ServerBootstrap`
+is called, a new listening socket is created to handle new inbound TCP connections. Let's take a look
+at the new `NIOAsyncChannel` based bind methods.
+
+```swift
+let serverChannel = try await ServerBootstrap(group: eventLoopGroup)
+    .bind(
+        host: "127.0.0.1",
+        port: 0,
+        childChannelInboundType: ByteBuffer.self,
+        childChannelOutboundType: ByteBuffer.self
+    )
+
+    try await withThrowingDiscardingTaskGroup { group in
+        for try await connectionChannel in serverChannel.inboundStream {
+            group.addTask {
+                do {
+                    for try await inboundData in connectionChannel.inboundStream {
+                        try await  connectionChannel.outboundWriter.write(inboundData)
+                    }
+                } catch {
+                    // Handle errors
+                }
+            }
+        }
+    }
+```
+
+In the above code, we are bootstrapping a new TCP server which we assign to `serverChannel`. The type of 
+`serverChannel` is a `NIOAsyncChannel` of `NIOAsyncChannel`. This is due to the fact that each
+inbound connection gets its own separate child channel.
+Afterward, we are handling each inbound connection in a separate child task and echo the data back.
+
+> Important: Make sure to use discarding task groups which are automatically reaping finished child tasks.
+Normal task groups will result in a memory leak since they do not reap their child tasks automatically.
+
+The above works create for channels that do not have protocol negotiation handlers in their pipeline
+since we can define the child channel's inbound and outbound type when calling `bind()`. In the case
+of protocol negotiation, we have to use a different bind method since now the types of the child channels
+are only known once protocol negotiation is done. Let's walk through how to setup a `ServerBootstrap`
+with protocol negotiation.
+
+First, we have to define our negotiation result. For this example, we are negotiating between a
+`String` based and `UInt8` based channel. Additionally, we also need an error that we can throw
```

Comment:
Would be nice if we could use a real example here; e.g. HTTP1 vs HTTP2

Reply:
Agreed, but we don't have those yet. Might be worth to swing around later

Reply:
I think that's probably fine; `AsyncChannel` is still SPI and we're going to do H2 soon.

Reply:
Discussed offline: Since we do not know the type that HTTP2 will use for its async abstraction we decided not to showcase it here.